File: products/opensyber/apps/api/src/auth/oauth.ts
/**
 * OAuth2 provider implementation for Google and GitHub.
 * Creates authorization URLs and exchanges authorization codes for access tokens.
 */

import type { OAuth2Config, OAuth2Provider, AuthUser } from './types.js';

export class OAuth2Error extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'OAuth2Error';
  }
}

export class GoogleOAuth2Provider implements OAuth2Provider {
  private config: OAuth2Config;
  private authUrl = 'https://accounts.google.com/o/oauth2/v2/auth';
  private tokenUrl = 'https://oauth2.googleapis.com/token';
  private userInfoUrl = 'https://www.googleapis.com/oauth2/v1/userinfo';

  constructor(config: OAuth2Config) {
    if (config.provider !== 'google') throw new OAuth2Error('Invalid provider');
    this.config = config;
  }

  getAuthURL(state: string): string {
    const params = new URLSearchParams({
      client_id: this.config.clientId,
      redirect_uri: this.config.redirectUri,
      response_type: 'code',
      scope: 'openid email profile',
      state,
    });
    return `${this.authUrl}?${params.toString()}`;
  }

  async exchangeCode(code: string): Promise<{ accessToken: string; user: Partial<AuthUser> }> {
    const body = new URLSearchParams({
      client_id: this.config.clientId,
      client_secret: this.config.clientSecret,
      code,
      grant_type: 'authorization_code',
      redirect_uri: this.config.redirectUri,
    });

    const tokenRes = await fetch(this.tokenUrl, { method: 'POST', body });
    if (!tokenRes.ok) throw new OAuth2Error('Token exchange failed');

    const { access_token } = (await tokenRes.json()) as { access_token: string };

    const userRes = await fetch(this.userInfoUrl, {
      headers: { Authorization: `Bearer ${access_token}` },
    });
    if (!userRes.ok) throw new OAuth2Error('Failed to fetch user info');

    const userInfo = (await userRes.json()) as { email: string; name: string };
    return {
      accessToken: access_token,
      user: { email: userInfo.email, name: userInfo.name, role: 'user' },
    };
  }
}

export class GitHubOAuth2Provider implements OAuth2Provider {
  private config: OAuth2Config;
  private authUrl = 'https://github.com/login/oauth/authorize';
  private tokenUrl = 'https://github.com/login/oauth/access_token';
  private userInfoUrl = 'https://api.github.com/user';

  constructor(config: OAuth2Config) {
    if (config.provider !== 'github') throw new OAuth2Error('Invalid provider');
    this.config = config;
  }

  getAuthURL(state: string): string {
    const params = new URLSearchParams({
      client_id: this.config.clientId,
      redirect_uri: this.config.redirectUri,
      scope: 'read:user user:email',
      state,
    });
    return `${this.authUrl}?${params.toString()}`;
  }

  async exchangeCode(code: string): Promise<{ accessToken: string; user: Partial<AuthUser> }> {
    const body = new URLSearchParams({
      client_id: this.config.clientId,
      client_secret: this.config.clientSecret,
      code,
    });

    const tokenRes = await fetch(this.tokenUrl, {
      method: 'POST',
      body,
      headers: { Accept: 'application/json' },
    });
    if (!tokenRes.ok) throw new OAuth2Error('Token exchange failed');

    const { access_token } = (await tokenRes.json()) as { access_token: string };

    const userRes = await fetch(this.userInfoUrl, {
      headers: { Authorization: `Bearer ${access_token}` },
    });
    if (!userRes.ok) throw new OAuth2Error('Failed to fetch user info');

    const userInfo = (await userRes.json()) as { email: string; name: string };
    return {
      accessToken: access_token,
      user: { email: userInfo.email, name: userInfo.name, role: 'user' },
    };
  }
}

export function createOAuth2Provider(config: OAuth2Config): OAuth2Provider {
  if (config.provider === 'google') return new GoogleOAuth2Provider(config);
  if (config.provider === 'github') return new GitHubOAuth2Provider(config);
  throw new OAuth2Error(`Unsupported provider: ${config.provider}`);
}

File: products/amliq/api/internal/config/oauth.go
package config

import "os"

type OAuthProvider struct {
	ClientID     string
	ClientSecret string
	AuthURL      string
	TokenURL     string
	UserInfoURL  string
	Scopes       []string
}

type OAuthConfig struct {
	FrontendURL string
	APIURL      string
	Providers   map[string]OAuthProvider
}

func LoadOAuth() OAuthConfig {
	cfg := OAuthConfig{
		FrontendURL: envOr("FRONTEND_URL", "http://localhost:5173"),
		APIURL:      envOr("API_URL", "http://localhost:8080"),
		Providers:   make(map[string]OAuthProvider),
	}
	if id := os.Getenv("GOOGLE_CLIENT_ID"); id != "" {
		cfg.Providers["google"] = OAuthProvider{
			ClientID: id, ClientSecret: os.Getenv("GOOGLE_CLIENT_SECRET"),
			AuthURL:     "https://accounts.google.com/o/oauth2/v2/auth",
			TokenURL:    "https://oauth2.googleapis.com/token",
			UserInfoURL: "https://www.googleapis.com/oauth2/v2/userinfo",
			Scopes:      []string{"openid", "email", "profile"},
		}
	}
	if id := os.Getenv("GITHUB_CLIENT_ID"); id != "" {
		cfg.Providers["github"] = OAuthProvider{
			ClientID: id, ClientSecret: os.Getenv("GITHUB_CLIENT_SECRET"),
			AuthURL:     "https://github.com/login/oauth/authorize",
			TokenURL:    "https://github.com/login/oauth/access_token",
			UserInfoURL: "https://api.github.com/user",
			Scopes:      []string{"user:email"},
		}
	}
	if id := os.Getenv("MICROSOFT_CLIENT_ID"); id != "" {
		cfg.Providers["microsoft"] = OAuthProvider{
			ClientID: id, ClientSecret: os.Getenv("MICROSOFT_CLIENT_SECRET"),
			AuthURL:     "https://login.microsoftonline.com/common/oauth2/v2.0/authorize",
			TokenURL:    "https://login.microsoftonline.com/common/oauth2/v2.0/token",
			UserInfoURL: "https://graph.microsoft.com/v1.0/me",
			Scopes:      []string{"openid", "email", "profile", "User.Read"},
		}
	}
	if id := os.Getenv("LINKEDIN_CLIENT_ID"); id != "" {
		cfg.Providers["linkedin"] = OAuthProvider{
			ClientID: id, ClientSecret: os.Getenv("LINKEDIN_CLIENT_SECRET"),
			AuthURL:     "https://www.linkedin.com/oauth/v2/authorization",
			TokenURL:    "https://www.linkedin.com/oauth/v2/accessToken",
			UserInfoURL: "https://api.linkedin.com/v2/userinfo",
			Scopes:      []string{"openid", "email", "profile"},
		}
	}
	return cfg
}

func envOr(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}
